mantidproject/mantid
# Inelastic Support for MonteCarloAbsorption algorithm

This work adds support for inelastic instruments to the `MonteCarloAbsorption` algorithm. I have chosen not to include this in the interface and require it to have been set on the workspace. Issue #15922 has been created to ensure `ConvertUnits` does this consistently for all modes.

**To test:**

The following script runs the MC absorption calculation against the flat plate & cylinder algorithms that use numerical integration. Plotting both against each other should show similar factors calculated in each case.

``` python
from mantid.simpleapi import *

# -------------------------------------
# Comment/uncomment the required mode
#emode = 'Indirect'
emode = 'Direct'

if emode == 'Indirect':
    sample_file = 'IRS26176.RAW'
    efixed = 1.845
else:
    sample_file = 'MAR11060.RAW'
    efixed = 12.0

# Vanadium
sample_formula = 'V'
# geometry in cm
sample_height = 3.0
sample_radius = 0.1
# atoms/A^3
sample_density = 0.073

# simulation options
nlambda_cyl = 20
nlambda_mc = 20

# ------------------- Setup the workspace ------------------------

red_ws = Load(Filename=sample_file)
# Just do a single spectrum
red_ws = CropWorkspace(red_ws, StartWorkspaceIndex=2, EndWorkspaceIndex=2)
SetInstrumentParameter(red_ws, ParameterName="deltaE-mode", Value=emode, ParameterType="String")
if emode == 'Indirect':
    SetInstrumentParameter(red_ws, ParameterName="Efixed",Value=str(efixed), ParameterType="Number")
elif emode == 'Direct':
    AddSampleLog(red_ws, LogName='Ei', LogText=str(efixed), LogType='Number')
red_ws = ConvertUnits(red_ws, Target='Wavelength',
                      EMode=emode)
SetSampleMaterial(red_ws, ChemicalFormula=sample_formula,
                  SampleNumberDensity=sample_density)

# --------------------- Cylinder absorption ---------------------

red_ws_ass_cylabs = CylinderAbsorption(InputWorkspace=red_ws,
                                       NumberOfWavelengthPoints=nlambda_cyl,
                                       CylinderSampleHeight=sample_height,
                                       CylinderSampleRadius=sample_radius,
                                       NumberOfSlices=100,
                                       NumberOfAnnuli=100,
                                       EMode=emode, EFixed=efixed)

# --------------------- Flat Plate absorption ---------------------

red_ws_ass_plate = FlatPlateAbsorption(InputWorkspace=red_ws,
                                       NumberOfWavelengthPoints=nlambda_cyl,
                                       SampleHeight=sample_height,
                                       SampleWidth=sample_radius,
                                       SampleThickness=sample_radius,
                                       ElementSize=0.1,
                                       EMode=emode, EFixed=efixed)

# --------------------- Monte Carlo absorption cylinder ---------------------

cylinder_template = '''<cylinder id="A">
  <centre-of-bottom-base x="0.0" y="{}" z="0.0" />
  <axis x="0.0" y="1.0" z="0.0" />
  <radius val="{}" />
  <height val="{}" />
</cylinder>'''
shape_xml = cylinder_template.format(-0.5*sample_height*0.01,sample_radius*0.01, sample_height*0.01)
CreateSampleShape(red_ws, ShapeXML=shape_xml)
SetSampleMaterial(red_ws, ChemicalFormula=sample_formula,
                  SampleNumberDensity=sample_density)
red_ws_ass_cyl_mc = MonteCarloAbsorption(InputWorkspace=red_ws,
                                     NumberOfWavelengthPoints=nlambda_mc,
                                         EventsPerPoint=10000)


# --------------------- Monte Carlo absorption flatplate ---------------------
szX, szY, szZ = 0.5*sample_radius*0.01, 0.5*sample_height*0.01, 0.5*sample_radius*0.01
plate_template = '''<cuboid id="A">
  <left-front-bottom-point x="{}" y="{}" z="{}"  />
  <left-front-top-point  x="{}" y="{}" z="{}"  />
  <left-back-bottom-point  x="{}" y="{}" z="{}"  />
  <right-front-bottom-point  x="{}" y="{}" z="{}"  />
</cuboid>'''
shape_xml = plate_template.format(szX, -szY, -szZ,
                                  szX, szY, -szZ,
                                  szX, -szY, szZ,
                                  -szX, -szY, -szZ)
CreateSampleShape(red_ws, ShapeXML=shape_xml)
SetSampleMaterial(red_ws, ChemicalFormula=sample_formula,
                  SampleNumberDensity=sample_density)
red_ws_ass_plate_mc = MonteCarloAbsorption(InputWorkspace=red_ws,
                                           NumberOfWavelengthPoints=nlambda_mc,
                                           EventsPerPoint=10000)
```

Fixes #15770

[Release Notes](https://github.com/mantidproject/mantid/blob/1845a6b831ba124f4dfcbd250b983971a2cd09b5/docs/source/release/v3.7.0/framework.rst#improved)

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
